e2e/js_image/runfiles.bzl

```diff
@@ -0,0 +1,129 @@
+        content = _LAUNCHER_TMPL.format(executable_path = executable_path),
+        is_executable = True,
+    )
+
+    return launcher
+
+def _runfiles_impl(ctx):
+    default = ctx.attr.binary[DefaultInfo]
+
+    executable = default.files_to_run.executable
+    executable_path = "/".join([ctx.attr.root, executable.short_path])
+    original_executable_path = executable_path.replace(".sh", "_.sh")
+    launcher = _write_laucher(ctx, original_executable_path)
+
+    files = depset(transitive = [default.files, default.default_runfiles.files])
+    file_map = {
+        original_executable_path: executable,
+        executable_path: launcher,
+    }
+
+    manifest = default.files_to_run.runfiles_manifest
+    runfiles_dir = "/".join([ctx.attr.root, manifest.short_path.replace(manifest.basename, "")[:-1]])
+
+    for file in files.to_list():
+        file_map[_runfile_path(ctx, file, runfiles_dir)] = file
+
+    # executable and launcher should not go into runfiles directory so we add it to files here
+    files = depset([executable, launcher], transitive = [files])
+
+    symlinks = []
+
+    # NOTE: pkg_tar is not capable of handling relative symlinks so they always have to be absolute
+    # NOTE: symlinks is different than root_symlinks. See: https://bazel.build/rules/rules#runfiles_symlinks for distinction between root_symlinks and symlinks
+    # and why they have to be handled differently.
+    for symlink in default.data_runfiles.symlinks.to_list():
+        destination = "/".join([runfiles_dir, ctx.workspace_name, symlink.path])
+        if file_map[destination]:
+            file_map.pop(destination)
+        info = PackageSymlinkInfo(
+            target = "/%s" % _runfile_path(ctx, symlink.target_file, runfiles_dir),
+            destination = destination,
+            attributes = {"mode": "0777"},
+        )
+        symlinks.append([info, symlink.target_file.owner])
+
+    for symlink in default.data_runfiles.root_symlinks.to_list():
+        destination = "/".join([runfiles_dir, symlink.path])
+        if file_map[destination]:
+            file_map.pop(destination)
+        info = PackageSymlinkInfo(
+            target = "/%s" % _runfile_path(ctx, symlink.target_file, runfiles_dir),
+            destination = destination,
+            attributes = {"mode": "0777"},
+        )
+        symlinks.append([info, symlink.target_file.owner])
+
+    return [
+        PackageFilegroupInfo(
+            pkg_dirs = [],
+            pkg_files = [
+                [PackageFilesInfo(
+                    dest_src_map = file_map,
+                    attributes = {},
+                ), ctx.label],
+            ],
+            pkg_symlinks = symlinks,
+        ),
+        DefaultInfo(files = files),
+    ]
+
+expand_runfiles = rule(
+    implementation = _runfiles_impl,
+    attrs = {
+        "binary": attr.label(mandatory = True),
+        "root": attr.string(),
+    },
+)
+
+def js_image_layer(name, binary, root = None, **kwargs):
```

Comment:
todo to promote this from userspace to the rules_js public API?

Reply:
I am not sure which way would be the best to go. if we expose this as public then we'd have to be really careful to not introduce any breaking changes which would be hard to do since users could rely on the structure of the runfiles laid out inside the container image. 

maybe it should stay as userland for a while? at least till we get enough feedback?